lib/watchers/fs-watcher.js

```diff
@@ -0,0 +1,439 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2015, Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var fs = require('fs');
+var vasync = require('vasync');
+
+// when to retry if startup fails (in ms)
+var retryDelay = 10000;
+
+
+function FsWatcher(opts) {
+    var self = this;
+
+    assert.object(opts, 'opts');
+    assert.object(opts.log, 'opts.log');
+    assert.func(opts.updateVm, 'opts.updateVm');
+
+    // Yay bunyan!
+    self.log = opts.log.child({watcher: 'fs-watcher'});
+
+    self.updateVm = opts.updateVm;
+
+    self.configWatchers = {};
+
+    self.isDirty = false;
+    self.isProcessing = false;
+
+    // lastSeenVms will contain a map of uuid -> timestamp with the latest
+    // timestamp seen for any of:
+    //
+    //  /etc/zones/<uuid>.xml
+    //  /zones/<uuid>/config/metadata.json
+    //  /zones/<uuid>/config/routes.json
+    //  /zones/<uuid>/config/tags.json
+    //
+    // It gets populated initially when we first load the VM data.
+    self.lastSeenVms = null;
+    self.zonesWatcher = null;
+}
+
+function findVmUuidFromFilename(filename) {
+    var match;
+    var re;
+    var vm = null;
+    var xmlFilenameLength = 40;
+
+    re = /(^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}).xml$/;
+
+    if (filename && filename.length === xmlFilenameLength) {
+        match = filename.match(re);
+
+        if (match) {
+            vm = match[1];
+        }
+    }
+
+    return (vm);
+}
+
+/*
+ * Finds the last modified timestamp for a VM via /etc/zones/<uuid>.xml and
```

Comment:
I would suggest changing `for a VM` to `for VMs`, otherwise it's a bit confusing whether this deals with one or several VMs.